modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/planner/PlannerTimeoutTest.java

```diff
@@ -0,0 +1,157 @@
+    /** */
+    private static final long PLANNER_TIMEOUT = 1_000;
+
+    /** */
+    private static final IgniteTypeFactory typeFactory = new IgniteTypeFactory(IgniteTypeSystem.INSTANCE);
+
+    /** */
+    @Test
+    public void testLongPlanningTimeout() throws Exception {
+        SchemaPlus schema = createSchema("PUBLIC",
+            Pair.of("T1", createTestTable(Pair.of("A", Integer.class), Pair.of("B", Integer.class))),
+            Pair.of("T2", createTestTable(Pair.of("A", Integer.class), Pair.of("C", Integer.class)))
+        );
+
+        String sql = "SELECT * FROM T1 JOIN T2 ON T1.A = T2.A";
+
+        PlanningContext ctx = PlanningContext.builder()
+            .parentContext(BaseQueryContext.builder()
+                .frameworkConfig(newConfigBuilder(FRAMEWORK_CONFIG)
+                    .defaultSchema(schema)
+                    .costFactory(new IgniteCostFactory(1, 100, 1, 1))
+                    .build())
+                .logger(log)
+                .build()
+            )
+            .plannerTimeout(PLANNER_TIMEOUT)
+            .query(sql)
+            .build();
+
+        AtomicReference<IgniteRel> plan = new AtomicReference<>();
+        AtomicReference<RelOptPlanner.CannotPlanException> plannerError = new AtomicReference<>();
+
+        GridTestUtils.assertTimeout(3 * PLANNER_TIMEOUT, TimeUnit.MILLISECONDS, () -> {
+            try (IgnitePlanner planner = ctx.planner()) {
+                plan.set(physicalPlan(planner, ctx.query()));
+
+                VolcanoPlanner volcanoPlanner = GridTestUtils.getFieldValue(planner, "planner");
+
+                assertNotNull(volcanoPlanner);
+
+                GridTestUtils.assertThrowsWithCause(volcanoPlanner::checkCancel, VolcanoTimeoutException.class);
+            }
+            catch (RelOptPlanner.CannotPlanException e) {
+                plannerError.set(e);
+            }
+            catch (Exception e) {
+                throw new RuntimeException("Planning failed", e);
+            }
+        });
+
+        assertTrue(plan.get() != null || plannerError.get() != null);
+
+        if (plan.get() != null) {
+            new RelVisitor() {
+                @Override public void visit(
+                    RelNode node,
+                    int ordinal,
+                    @Nullable RelNode parent) {
+                    assertNotNull(node.getTraitSet().getTrait(IgniteConvention.INSTANCE.getTraitDef()));
+                    super.visit(node, ordinal, parent);
+                }
+            }.go(plan.get());
+        }
+    }
+
+    /** */
+    private static SchemaPlus createSchema(String name, Pair<String, TestTable>... tables) {
+        IgniteSchema schema = new IgniteSchema(name);
+
+        for (Pair<String, TestTable> tbl: tables)
+            schema.addTable(tbl.left, tbl.right);
+
+        return createRootSchema(false).add(name, schema);
+    }
+
+    /** */
+    private static TestTable createTestTable(Pair<String, Class<?>>... cols) {
```

Comment:
There are already `createSchema` and `createTable` (with distribution and columns) methods exists in `AbstractPlannerTest`, you should reuse these methods.